src/main/java/com/minecolonies/coremod/util/InventoryUtils.java

```diff
@@ -294,603 +363,1248 @@ public static boolean hasToolLevel(final String tool, final ItemStack stack, fin
+     * to the given target {@link ICapabilityProvider}.
+     *
+     * @param sourceHandler  The {@link IItemHandler} that works as Source.
+     * @param sourceIndex    The index of the slot that is being extracted
+     *                       from.
+     * @param targetProvider The {@link ICapabilityProvider} that works as
+     *                       Target.
+     * @return True when the swap was successful, false when not.
+     */
+    public static boolean transferItemStackIntoNextFreeSlotInProvider(
+                                                                       @NotNull final IItemHandler sourceHandler,
+                                                                       @NotNull int sourceIndex,
+                                                                       @NotNull ICapabilityProvider targetProvider)
+    {
+        for (IItemHandler handler : getItemHandlersFromProvider(targetProvider))
+        {
+            if (transferItemStackIntoNextFreeSlotInItemHandlers(sourceHandler, sourceIndex, handler))
             {
-                slot = getOpenSlot(inventory);
+                return true;
             }
+        }
 
-            if (slot < 0)
+        return false;
+    }
+
+    /**
+     * Method to swap the ItemStacks from the given source {@link IItemHandler}
+     * to the given target {@link IItemHandler}.
+     *
+     * @param sourceHandler The {@link IItemHandler} that works as Source.
+     * @param sourceIndex   The index of the slot that is being extracted from.
+     * @param targetHandler The {@link IItemHandler} that works as Target.
+     * @return True when the swap was successful, false when not.
+     */
+    public static boolean transferItemStackIntoNextFreeSlotInItemHandlers(@NotNull final IItemHandler sourceHandler, @NotNull int sourceIndex, @NotNull IItemHandler targetHandler)
+    {
+        ItemStack sourceStack = sourceHandler.extractItem(sourceIndex, Integer.MAX_VALUE, true);
+        final ItemStack originalStack = sourceStack.copy();
+
+        for (int i = 0; i < targetHandler.getSlots(); i++)
+        {
+            sourceStack = targetHandler.insertItem(i, sourceStack, false);
+            if (isItemStackEmpty(sourceStack))
             {
-                return stackSize;
+                sourceHandler.extractItem(sourceIndex, Integer.MAX_VALUE, false);
+                return true;
             }
-            else
+        }
+
+        if (!ItemStack.areItemStacksEqual(sourceStack, originalStack) && compareItemStacksIgnoreStackSize(sourceStack, originalStack))
+        {
+            int usedAmount = getItemStackSize(sourceStack) - getItemStackSize(originalStack);
+            sourceHandler.extractItem(sourceIndex, usedAmount, false);
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Method to compare to stacks, ignoring their stacksize.
+     *
+     * @param itemStack1 The left stack to compare.
+     * @param itemStack2 The ri
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Introduce a new variable instead of reusing the parameter "itemStack". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1226)